### PR TITLE
added csv download link to budget_investments (issue 1625)

### PR DESCRIPTION
**Pull request by raul-fuentes**
Where
=====
* **Related Issue:**  https://github.com/consul/consul/issues/1625

What
====
- Added download csv link to budget investments

How
===
- it just replicates the content of the html table into CSV

Screenshots
===========
screenshot from 2017-11-21 13 35 26

Test
====
- works with current tests

Deployment
==========
- No

Warnings
========
- No

### CODE REVIEW

File: app/controllers/admin/budget_investments_controller.rb [hunk 1]
```diff
@@ -43,6 +52,41 @@ def toggle_selection
...
+                          else
+                            t("admin.budget_investments.index.no_admin_assigned")
+                          end
+          valuators = if investment.valuators.empty?
+                        t("admin.budget_investments.index.no_valuators_assigned")
+                      else
+                        investment.valuators.collect(&:description_or_name).join(', ')
+                      end
+          heading_name = investment.heading.name
+          price = t("admin.budget_investments.index.feasibility.#{investment.feasibility}", price: investment.formatted_price)
```
Comment (bertocq) [inline app/controllers/admin/budget_investments_controller.rb]: Could we do the interpolation outside of the `i18n.t` method call? So i18n-tasks will be able to check this string is in use?

File: app/views/admin/budget_investments/_investments.html.erb [hunk 2]
```diff
@@ -55,26 +58,9 @@
...
-          <%= link_to toggle_selection_admin_budget_budget_investment_path(@budget,
-                                                                           investment,
-                                                                           filter: params[:filter],
-                                                                           page: params[:page]),
-                      method: :patch,
-                      remote: true,
-                      class: "button small hollow expanded" do %>
-            <%= t("admin.budget_investments.index.select") %>
-          <% end %>
+        <%= options = investment_selected_link_options(investment) %>
```
Comment (bertocq) [inline app/views/admin/budget_investments/_investments.html.erb]: This `<%=` should be `<%` to avoid printing the result of the function. We normally require screenshots to both improve the review and also to "force" the coder to manually verify interface changes just in case something happens...

File: app/views/admin/valuators/search.html.erb [hunk 3]
```diff
@@ -29,7 +29,12 @@
                 <%= t("admin.valuators.index.no_description") %>
               <% end %>
             <td>
-              <% unless user.valuator? %>
+              <% if user.valuator? %>
+                <%= link_to t("admin.valuators.valuator.delete"),
```
Comment (bertocq) [inline app/views/admin/valuators/search.html.erb]: Not sure why this changes appear here, seems like your branch isn't rebased with master, and some commits from angel and xuanxu came along on the PR making it a bit more difficult to review

File: spec/features/admin/budget_investments_spec.rb [hunk 4]
```diff
@@ -588,7 +588,23 @@
...
+      investment2 = create(:budget_investment, :feasible, budget: @budget)
 
+      visit admin_budget_budget_investments_path(@budget, format: :csv)
+
+      header = page.response_headers['Content-Disposition']
+      header.should match(/^attachment/)
+      header.should match(/filename="budget_investments.csv"$/)
+
+      expect(page).to have_content investment2.title
+      expect(page).to have_content investment1.title
```
Comment (bertocq) [inline spec/features/admin/budget_investments_spec.rb]: I like this way to check de csv download 👍  😃  I'm only missing a negative scenario, I mean: would it be possible to check that filtering by feasible only budget investments, the unfeasible ones doesn't appear at the file? Or any other scenario that you see easier to reproducer, where we can verify that only the investments that the user see are at the html view the ones exported in the file :)

File: app/controllers/admin/budget_investments_controller.rb (hunk 1 above)
Comment (rfuentesm2c) [reply app/controllers/admin/budget_investments_controller.rb]: In admin.yml, the translation admin.budget_investments.index.feasibility includes at the end (%{price})".
Should I change the yml file?

File: app/views/admin/valuators/search.html.erb (hunk 3 above)
Comment (rfuentesm2c) [reply app/views/admin/valuators/search.html.erb]: Im working on it

File: app/controllers/admin/budget_investments_controller.rb (hunk 1 above)
Comment (bertocq) [reply app/controllers/admin/budget_investments_controller.rb]: No, what I mean is that instead of:

File: spec/features/admin/budget_investments_spec.rb (hunk 4 above)
Comment (rfuentesm2c) [reply spec/features/admin/budget_investments_spec.rb]: Ok, It will be in a moment

File: app/controllers/admin/budget_investments_controller.rb (hunk 1 above)
Comment (rfuentesm2c) [reply app/controllers/admin/budget_investments_controller.rb]: Oh, you're right. I'll change it

File: app/controllers/admin/budget_investments_controller.rb [hunk 5]
```diff
@@ -43,6 +52,42 @@ def toggle_selection
 
   private
 
+    def set_csv
```
Comment (voodoorai2000) [inline app/controllers/admin/budget_investments_controller.rb]: This method is a bit too large and does not belong in a controller
Could you extract this logic to the investment's model?
Something on these lines http://railscasts.com/episodes/362-exporting-csv-and-excel

File: app/helpers/budgets_helper.rb [hunk 6]
```diff
@@ -1,5 +1,23 @@
 module BudgetsHelper
 
+  def csv_params
+    csv_params = params.clone.merge(format: :csv)
+    csv_params.delete(:page)
+    csv_params
+  end
+
+  def investment_selected_link_options(investment)
```
Comment (voodoorai2000) [inline app/helpers/budgets_helper.rb]: Great refactoring 👍 
Keep in mind that if you run into a refactoring to do, it's nicer to do it in a separate PR

File: app/views/admin/budget_investments/_investments.html.erb [hunk 7]
```diff
@@ -55,26 +58,9 @@
...
-                                                                           investment,
-                                                                           filter: params[:filter],
-                                                                           page: params[:page]),
-                      method: :patch,
-                      remote: true,
-                      class: "button small hollow expanded" do %>
-            <%= t("admin.budget_investments.index.select") %>
-          <% end %>
+        <% options = investment_selected_link_options(investment) %>
+        <%= link_to toggle_selection_admin_budget_budget_investment_path(@budget,  investment, filter: params[:filter], page: params[:page]), method: :patch, remote: true, class: options[:link_class] do %>
```
Comment (voodoorai2000) [inline app/views/admin/budget_investments/_investments.html.erb]: This line is too long
Try to keep the length of lines to 80 characters or less :ok_hand:
A helper method would be great here or just splitting attributes in separate lines

File: spec/features/admin/budget_investments_spec.rb [hunk 8]
```diff
@@ -588,7 +588,36 @@
         expect(page).to_not have_link('Selected')
       end
     end
+  end
+
+  context "Selecting csv" do
```
Comment (voodoorai2000) [inline spec/features/admin/budget_investments_spec.rb]: Great 👌 
Please also test for the presence of each attribute in the csv

**Comment by raul-fuentes** [comment]
hello @voodoorai2000  the changes that you requested are already made, I hope that now everything is correct.

**Comment by voodoorai2000** [comment]
Excellent @raul-fuentes!
